### ISSUE TITLE
Propose extending core optimizer and create_hooks to support Horovod

### ISSUE THREAD

**Issue by cwbeitel** [tensorflow/tensor2tensor#937]
In the style of `SyncReplicasOptimizer`, using Horovod for managing the syncing of gradients requires both wrapping the optimizer and a custom hook, see the example here: https://github.com/uber/horovod

Including the custom hook could be accomplished by overriding [train_hooks](https://github.com/tensorflow/tensor2tensor/blob/3b5d95c6e2327017e06fe721d2810579f7917320/tensor2tensor/utils/t2t_model.py#L130) in T2TModel,
e.g.

```python
import horovod as hvd
  
class Img2imgTransformerHVD(Img2imgTransformer):
  
  @staticmethod
  def train_hooks():
    """Add Horovod hook to perform gradiant sync."""
    # Should be added conditionally depending on hparams.

    hvd.init()
    return [hvd.BroadcastGlobalVariablesHook(0)]

```

But it would be awkward to need to subclass every model you might want to use in this way.

Alternatively, the additional train hook could be added in [create_hooks](https://github.com/tensorflow/tensor2tensor/blob/3b5d95c6e2327017e06fe721d2810579f7917320/tensor2tensor/utils/trainer_lib.py#L228) based on whether --use_hvd is specified.

With this approach, `use_hvd` can also be passed to [optimize](https://github.com/tensorflow/tensor2tensor/blob/79deef4b334d03d371c9d5033cfb99979b304ad7/tensor2tensor/utils/optimize.py#L42)  in the same manner as `use_tpu` to determine whether to wrap the optimizer with `hvd.DistributedOptimizer`.

```python
def optimize(loss, learning_rate, hparams, use_tpu=False, use_hvd=False):
  """Minimize loss."""

  ...

  opt = ConditionalOptimizer(hparams.optimizer, learning_rate, hparams, use_tpu)
  if use_tpu:
    opt = tf.contrib.tpu.CrossShardOptimizer(opt)
  elif use_hvd:
    opt = hvd.DistributedOptimizer(opt)
```

/cc @rsepassi

**Comment by vsuthichai** [comment]
I'd be interested in seeing this happen as well.  I'm currently attempting to benchmark the transformer on 32 gpus across 4 machines, ie. 8 gpus each.

**Comment by billy19murahmi** [comment]
@cwbeitel Have you tried to implement it locally? 
I just started to experiment with t2t. The code kinda looks confusing as I'm trying to figure out where all to modify to add the horovod functionality.
Any pointers would be helpful.

**Comment by cwbeitel** [comment]
No but the two approaches I gave above should be good starting points if you want to give it a try!

The question I meant to raise by opening the issue was whether @rsepassi @lukaszkaiser et al. would consider this change to t2t in-scope given I was proposing making a change to the core (second case) instead of something in t2t_usr_dir (first case). If you want to continue making use of t2t improvements into the future it's probably best to incorporate a core change into the main repository instead of having to resolve merges every time you want to be up to date.

**Comment by EdwardZhang88** [comment]
Hi @cwbeitel @vsuthichai I have integrated Horovod v0.13.11 with tensor2tensor v1.8.0 (tensorflow v1.10) and managed to run distributed training over RoCE plus GPUDirect RDMA. 

However, when I re-run the the same distributed training task using plain old tensor2tensor v1.2.9 (tensorflow v1.4), it surprisingly achieves almost the same performance as the NCCL over RoCE one. The point is I didn't even specify any extra environment variables (NCCL and MPI mca parameters, etc.) like I did to run NCCL over RoCE. I kind of doubt if this is a normal case as I recall that to run tensorflow over RDMA + GDR, there are usually some prerequisites like [this](https://github.com/tensorflow/tensorflow/blob/9590c4c32dd4346ea5c35673336f5912c6072bf2/tensorflow/contrib/verbs/README.md) and [this](https://github.com/tensorflow/tensorflow/blob/9590c4c32dd4346ea5c35673336f5912c6072bf2/tensorflow/contrib/gdr/README.md). 

Have you compared the performance difference between the Horovod task and its original tensor2tensor one? 

BTW, the tensor2tensor distributed training appears to be really confusing and neither the [new document](https://github.com/tensorflow/tensor2tensor/blob/2250c2c2fa286b525760455f72f743fadf4fb45e/docs/distributed_training.md) nor the [old document ](https://github.com/tensorflow/tensor2tensor/blob/ed9e3bdfd0292d4b6e5b1a1bf272146c8e2f5e9f/docs/distributed_training.md) are clear enough, I have to somehow tweak the worker/master parameter to get it work using tensor2tensor v1.2.9 (tensorflow v1.4). Down below is the command I use. 


Does this really trigger the distributed async training as expected? If so, does it mean that the gRPC backend of Tensorflow is smart enough to automatically detect the Mellanox NIC and set up the RoCE and GDR all correctly without any manual configuration? Thanks.

````
#ps
export TF_CONFIG='{"cluster":{"ps":["tensorflow-cluster-mnist-gpu-nmt-1-ps-z4f8-0:3334"],"worker":["tensorflow-cluster-mnist-gpu-nmt-1-worker-z4f8-0:3333","tensorflow-cluster-mnist-gpu-nmt-1-worker-z4f8-1:3333"]},"task":{"type":"ps","index":0},"environment":"cloud"}'
t2t-trainer \
    --master=grpc://tensorflow-cluster-mnist-gpu-nmt-1-ps-z4f8-0:3334 \
    --schedule= run_std_server \
    --model=transformer \
    --hparams_set=my_problem_hparams \
    --problem=my_problem

#woker 0
export TF_CONFIG='{"cluster":{"ps":["tensorflow-cluster-mnist-gpu-nmt-1-ps-z4f8-0:3334"],"master":["tensorflow-cluster-mnist-gpu-nmt-1-worker-z4f8-0:3333","tensorflow-cluster-mnist-gpu-nmt-1-worker-z4f8-1:3333"]},"task":{"type":"master","index":0},"environment":"cloud"}'
t2t-trainer \
    --master=grpc://tensorflow-cluster-mnist-gpu-nmt-1-worker-z4f8-0:3333 \
    --ps_replicas=1 \
    --worker_replicas=2 \
    --worker_gpu=1 \
    --worker_id=1 \
    --ps_gpu=0 \
    --schedule=train \
    --worker_job='/job:master \
    --model=transformer \
    --hparams_set=my_problem_hparams \
    --problem=my_problem

#worker 1
export TF_CONFIG='{"cluster":{"ps":["tensorflow-cluster-mnist-gpu-nmt-1-ps-z4f8-0:3334"],"master":["tensorflow-cluster-mnist-gpu-nmt-1-worker-z4f8-0:3333","tensorflow-cluster-mnist-gpu-nmt-1-worker-z4f8-1:3333"]},"task":{"type":"master","index":1},"environment":"cloud"}'
t2t-trainer \
    --master=grpc://tensorflow-cluster-mnist-gpu-nmt-1-worker-z4f8-1:3333 \
    --ps_replicas=1 \
    --worker_replicas=2 \
    --worker_gpu=1 \
    --worker_id=1 \
    --ps_gpu=0 \
    --schedule=train \
    --worker_job='/job:master' \
    --model=transformer \
    --hparams_set=my_problem_hparams \
    --problem=my_problem
```

**Comment by vsuthichai** [comment]
@EdwardZhang88 I've integrated Horovod 0.13.11 with T2T as well, latest from master.  I've shelved 
 the parameter server for now because I didn't see significant throughput increase compared to Horovod.  However, I did spend over a week trying to get the parameter server technique up and working with sync training, before and after the distributed training documentation updates.  I couldn't get it running successfully, until I changed the job name from master to chief within the cluster environment json and an extra code fix to start tf.train.Server manually for the chief.

**Comment by cwbeitel** [comment]
@EdwardZhang88 @vsuthichai That's cool you both got distributed training working! I've only done synchronous training not async or using Horovod so can't help there.

One sticking point I ran into with distributed training (that is explicitly mentioned in the docs now) is that all your jobs need to share the same checkpoint dir.

I'm going to close this issue since we haven't heard a response. This discussion would probably be a great fit for the gitter channel or the mailing list!
